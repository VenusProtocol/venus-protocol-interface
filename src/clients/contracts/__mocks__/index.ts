export const getTokenContract = jest.fn();
export const useTokenContract = () => getTokenContract;

export const getTokenContractByAddress = jest.fn();
export const useTokenContractByAddress = () => getTokenContractByAddress;

export const getVTokenContract = jest.fn();
export const useVTokenContract = () => getVTokenContract;

export const getVaiUnitrollerContract = jest.fn();
export const useVaiUnitrollerContract = () => getVaiUnitrollerContract;

export const getVaiVaultContract = jest.fn();
export const useVaiVaultContract = () => getVaiVaultContract;

export const getXvsVaultContract = jest.fn();
export const useXvsVaultContract = () => getXvsVaultContract;

export const getXvsVaultProxyContract = jest.fn();
export const useXvsVaultProxyContract = () => getXvsVaultProxyContract;

export const getXvsVaultStoreContract = jest.fn();

export const getComptrollerContract = jest.fn();
export const useComptrollerContract = () => getComptrollerContract;

export const getPriceOracleContract = jest.fn();
export const usePriceOracleContract = () => getPriceOracleContract;

export const getInterestModelContract = jest.fn();
export const useInterestModelContract = () => getInterestModelContract;

export const getVenusLensContract = jest.fn();
export const useVenusLensContract = () => getVenusLensContract;

export const getGovernorBravoDelegateContract = jest.fn();
export const useGovernorBravoDelegateContract = () => getGovernorBravoDelegateContract;

export const getXvsVestingProxyContract = jest.fn();
export const useXvsVestingProxyContract = () => getXvsVestingProxyContract;

export const getVrtConverterProxyContract = jest.fn();
export const useVrtConverterProxyContract = () => getVrtConverterProxyContract;

export const getVrtVaultProxyContract = jest.fn();
export const useVrtVaultProxyContract = () => getVrtVaultProxyContract;
