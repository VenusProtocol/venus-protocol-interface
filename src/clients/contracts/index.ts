export * from './getters';
export * from './hooks';
