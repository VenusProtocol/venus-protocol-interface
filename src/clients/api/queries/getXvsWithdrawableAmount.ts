import BigNumber from 'bignumber.js';
import { XvsVesting } from 'types/contracts';

export interface IGetXvsWithdrawableAmountInput {
  xvsVestingContract: XvsVesting;
  accountAddress: string;
}

export interface IGetXvsWithdrawableAmountOutput {
  totalWithdrawableAmount: BigNumber;
  totalVestedAmount: BigNumber;
  totalWithdrawnAmount: BigNumber;
}

const getXvsWithdrawableAmount = async ({
  xvsVestingContract,
  accountAddress,
}: IGetXvsWithdrawableAmountInput): Promise<IGetXvsWithdrawableAmountOutput> => {
  const resp = await xvsVestingContract.methods.getWithdrawableAmount(accountAddress).call();
  return {
    totalWithdrawableAmount: new BigNumber(resp.totalWithdrawableAmount),
    totalVestedAmount: new BigNumber(resp.totalVestedAmount),
    totalWithdrawnAmount: new BigNumber(resp.totalWithdrawnAmount),
  };
};

export default getXvsWithdrawableAmount;
