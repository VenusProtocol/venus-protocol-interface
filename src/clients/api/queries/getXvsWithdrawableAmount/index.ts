import BigNumber from 'bignumber.js';

import { XvsVesting } from 'types/contracts';

export interface GetXvsWithdrawableAmountInput {
  xvsVestingContract: XvsVesting;
  accountAddress: string;
}

export interface GetXvsWithdrawableAmountOutput {
  totalWithdrawableAmount: BigNumber;
  totalVestedAmount: BigNumber;
  totalWithdrawnAmount: BigNumber;
}

const getXvsWithdrawableAmount = async ({
  xvsVestingContract,
  accountAddress,
}: GetXvsWithdrawableAmountInput): Promise<GetXvsWithdrawableAmountOutput> => {
  const resp = await xvsVestingContract.getWithdrawableAmount(accountAddress);
  return {
    totalWithdrawableAmount: new BigNumber(resp.totalWithdrawableAmount.toString()),
    totalVestedAmount: new BigNumber(resp.totalVestedAmount.toString()),
    totalWithdrawnAmount: new BigNumber(resp.totalWithdrawnAmount.toString()),
  };
};

export default getXvsWithdrawableAmount;
