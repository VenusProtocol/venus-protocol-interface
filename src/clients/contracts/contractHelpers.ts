import Web3 from 'web3';

import { getWeb3NoAccount } from 'clients/web3';
import bep20Abi from 'constants/contractAbis/bep20.json';
import comptrollerAbi from 'constants/contractAbis/comptroller.json';
import interestModelAbi from 'constants/contractAbis/interestModel.json';
import oracleAbi from 'constants/contractAbis/oracle.json';
import vaiTokenAbi from 'constants/contractAbis/vaiToken.json';
import vaiUnitrollerAbi from 'constants/contractAbis/vaiUnitroller.json';
import vaiVaultAbi from 'constants/contractAbis/vaiVault.json';
import xvsVaultStoreAbi from 'constants/contractAbis/xvsVaultStore.json';
import xvsVaultAbi from 'constants/contractAbis/xvsVault.json';
import vbepAbi from 'constants/contractAbis/vbep.json';
import vbnbAbi from 'constants/contractAbis/vbnb.json';
import xvsAbi from 'constants/contractAbis/xvs.json';
import venusLensAbi from 'constants/contractAbis/venusLens.json';
import governorBravoAbi from 'constants/contractAbis/governorBravoDelegate.json';
import xvsVestingAbi from 'constants/contractAbis/xvsVesting.json';
import vrtConverterAbi from 'constants/contractAbis/vrtConverter.json';
import vrtTokenAbi from 'constants/contractAbis/vrtToken.json';
import vrtVaultAbi from 'constants/contractAbis/vrtVault.json';

import getContractAddress from '../../utilities/getContractAddress';
import * as constants from '../../constants/contracts';

const getContract = (abi: $TSFixMe, address: $TSFixMe, web3Contract: Web3) => {
  const web3 = web3Contract ?? getWeb3NoAccount();
  return new web3.eth.Contract(abi, address);
};

export const getVaiTokenContract = (web3: Web3) =>
  getContract(vaiTokenAbi, getContractAddress('vaiToken'), web3);

export const getVaiUnitrollerContract = (web3: Web3) =>
  getContract(vaiUnitrollerAbi, getContractAddress('vaiUnitroller'), web3);

export const getVaiVaultContract = (web3: Web3) =>
  getContract(vaiVaultAbi, getContractAddress('vaiVault'), web3);

export const getXvsVaultContract = (web3: Web3) =>
  getContract(xvsVaultAbi, getContractAddress('xvsVault'), web3);

export const getXvsVaultProxyContract = (web3: Web3) =>
  getContract(xvsVaultAbi, getContractAddress('xvsVaultProxy'), web3);

export const getXvsVaultStoreContract = (web3: Web3) =>
  getContract(xvsVaultStoreAbi, getContractAddress('xvsVaultStore'), web3);

export const getTokenContract = (web3: Web3, name: $TSFixMe) =>
  getContract(name === 'xvs' ? xvsAbi : bep20Abi, constants.getToken(name).address, web3);

export const getTokenContractByAddress = (
  web3: Web3,

  address: $TSFixMe,
) => getContract(vaiTokenAbi, address, web3);

export const getVbepContract = (web3: Web3, name: $TSFixMe) =>
  getContract(name === 'bnb' ? vbnbAbi : vbepAbi, constants.getVbepToken(name).address, web3);

export const getComptrollerContract = (web3: Web3) =>
  getContract(comptrollerAbi, getContractAddress('comptroller'), web3);

export const getPriceOracleContract = (web3: Web3) =>
  getContract(oracleAbi, getContractAddress('oracle'), web3);

export const getInterestModelContract = (web3: Web3, address: $TSFixMe) =>
  getContract(interestModelAbi, address, web3);

export const getVenusLensContract = (web3: Web3) =>
  getContract(venusLensAbi, getContractAddress('venusLens'), web3);

export const getGovernorBravoContract = (web3: Web3) =>
  getContract(governorBravoAbi, getContractAddress('governorBravoDelegator'), web3);

// VRT conversion
export const getXvsVestingProxyContract = (web3: Web3) =>
  getContract(xvsVestingAbi, getContractAddress('xvsVestingProxy'), web3);

export const getVrtConverterProxyContract = (web3: Web3) =>
  getContract(vrtConverterAbi, getContractAddress('vrtConverterProxy'), web3);

export const getVrtTokenContract = (web3: Web3) =>
  getContract(vrtTokenAbi, getContractAddress('vrtToken'), web3);

// VRT vault
export const getVrtVaultProxyContract = (web3: Web3) =>
  getContract(vrtVaultAbi, getContractAddress('vrtVaultProxy'), web3);
