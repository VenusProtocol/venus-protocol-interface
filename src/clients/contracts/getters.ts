import Web3 from 'web3';
import { AbiItem } from 'web3-utils';

import { TokenId, VTokenId } from 'types';
import { getWeb3NoAccount } from 'clients/web3';
import bep20Abi from 'constants/contracts/abis/bep20.json';
import comptrollerAbi from 'constants/contracts/abis/comptroller.json';
import interestModelAbi from 'constants/contracts/abis/interestModel.json';
import oracleAbi from 'constants/contracts/abis/oracle.json';
import vaiTokenAbi from 'constants/contracts/abis/vaiToken.json';
import vaiUnitrollerAbi from 'constants/contracts/abis/vaiUnitroller.json';
import vaiVaultAbi from 'constants/contracts/abis/vaiVault.json';
import xvsVaultStoreAbi from 'constants/contracts/abis/xvsVaultStore.json';
import xvsVaultAbi from 'constants/contracts/abis/xvsVault.json';
import vBep20Abi from 'constants/contracts/abis/vBep20.json';
import vBnbTokenAbi from 'constants/contracts/abis/vBnbToken.json';
import xvsTokenAbi from 'constants/contracts/abis/xvsToken.json';
import venusLensAbi from 'constants/contracts/abis/venusLens.json';
import governorBravoDelegateAbi from 'constants/contracts/abis/governorBravoDelegate.json';
import xvsVestingAbi from 'constants/contracts/abis/xvsVesting.json';
import vrtConverterAbi from 'constants/contracts/abis/vrtConverter.json';
import vrtTokenAbi from 'constants/contracts/abis/vrtToken.json';
import vrtVaultAbi from 'constants/contracts/abis/vrtVault.json';
import {
  Bep20,
  Comptroller,
  InterestModel,
  Oracle,
  VaiVault,
  VaiUnitroller,
  XvsVault,
  XvsVaultStore,
  VenusLens,
  GovernorBravoDelegate,
  XvsVesting,
  VrtVault,
  VrtConverter,
} from 'types/contracts';
import { getContractAddress, getToken, getVBepToken } from 'utilities';
import { TokenContract, VTokenContract } from './types';

const getContract = <T>(abi: AbiItem | AbiItem[], address: string, web3Instance: Web3) => {
  const web3 = web3Instance ?? getWeb3NoAccount();
  return new web3.eth.Contract(abi, address) as unknown as T;
};

export const getTokenContract = <T extends TokenId>(tokenId: T, web3: Web3): TokenContract<T> => {
  const tokenAddress = getToken(tokenId).address;

  if (tokenId === 'xvs') {
    return getContract<TokenContract<T>>(xvsTokenAbi as AbiItem[], tokenAddress, web3);
  }

  if (tokenId === 'vai') {
    return getContract<TokenContract<T>>(vaiTokenAbi as AbiItem[], tokenAddress, web3);
  }

  if (tokenId === 'vrt') {
    return getContract<TokenContract<T>>(vrtTokenAbi as AbiItem[], tokenAddress, web3);
  }

  return getContract<TokenContract<T>>(bep20Abi as AbiItem[], tokenAddress, web3);
};

export const getTokenContractByAddress = (address: string, web3: Web3): Bep20 =>
  getContract(bep20Abi as AbiItem[], address, web3) as unknown as Bep20;

export const getVTokenContract = <T extends VTokenId>(
  tokenId: T,
  web3: Web3,
): VTokenContract<T> => {
  const vBepTokenAddress = getVBepToken(tokenId).address;

  if (tokenId === 'bnb') {
    return getContract(
      vBnbTokenAbi as AbiItem[],
      vBepTokenAddress,
      web3,
    ) as unknown as VTokenContract<T>;
  }

  return getContract(
    vBep20Abi as AbiItem[],
    vBepTokenAddress,
    web3,
  ) as unknown as VTokenContract<T>;
};

export const getVaiUnitrollerContract = (web3: Web3) =>
  getContract(
    vaiUnitrollerAbi as AbiItem[],
    getContractAddress('vaiUnitroller'),
    web3,
  ) as unknown as VaiUnitroller;

export const getVaiVaultContract = (web3: Web3) =>
  getContract(
    vaiVaultAbi as AbiItem[],
    getContractAddress('vaiVault'),
    web3,
  ) as unknown as VaiVault;

export const getXvsVaultContract = (web3: Web3) =>
  getContract(
    xvsVaultAbi as AbiItem[],
    getContractAddress('xvsVault'),
    web3,
  ) as unknown as XvsVault;

export const getXvsVaultProxyContract = (web3: Web3) =>
  getContract(
    xvsVaultAbi as AbiItem[],
    getContractAddress('xvsVaultProxy'),
    web3,
  ) as unknown as XvsVault;

export const getXvsVaultStoreContract = (web3: Web3) =>
  getContract(
    xvsVaultStoreAbi as AbiItem[],
    getContractAddress('xvsVaultStore'),
    web3,
  ) as unknown as XvsVaultStore;

export const getComptrollerContract = (web3: Web3) =>
  getContract(
    comptrollerAbi as AbiItem[],
    getContractAddress('comptroller'),
    web3,
  ) as unknown as Comptroller;

export const getPriceOracleContract = (web3: Web3) =>
  getContract(oracleAbi as AbiItem[], getContractAddress('oracle'), web3) as unknown as Oracle;

export const getInterestModelContract = (address: string, web3: Web3) =>
  getContract(interestModelAbi as AbiItem[], address, web3) as unknown as InterestModel;

export const getVenusLensContract = (web3: Web3) =>
  getContract(
    venusLensAbi as AbiItem[],
    getContractAddress('venusLens'),
    web3,
  ) as unknown as VenusLens;

export const getGovernorBravoDelegateContract = (web3: Web3) =>
  getContract(
    governorBravoDelegateAbi as AbiItem[],
    getContractAddress('governorBravoDelegator'),
    web3,
  ) as unknown as GovernorBravoDelegate;

// VRT conversion
export const getXvsVestingProxyContract = (web3: Web3) =>
  getContract(
    xvsVestingAbi as AbiItem[],
    getContractAddress('xvsVestingProxy'),
    web3,
  ) as unknown as XvsVesting;

export const getVrtConverterProxyContract = (web3: Web3) =>
  getContract(
    vrtConverterAbi as AbiItem[],
    getContractAddress('vrtConverterProxy'),
    web3,
  ) as unknown as VrtConverter;

// VRT vault
export const getVrtVaultProxyContract = (web3: Web3) =>
  getContract(
    vrtVaultAbi as AbiItem[],
    getContractAddress('vrtVaultProxy'),
    web3,
  ) as unknown as VrtVault;
