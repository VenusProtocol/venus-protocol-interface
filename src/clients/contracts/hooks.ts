import { useMemo } from 'react';

import { TokenId, VTokenId } from 'types';
import { useWeb3 } from 'clients/web3';
import {
  getComptrollerContract,
  getInterestModelContract,
  getPriceOracleContract,
  getTokenContract,
  getVaiUnitrollerContract,
  getVaiVaultContract,
  getVTokenContract,
  getVenusLensContract,
  getXvsVaultProxyContract,
  getXvsVaultContract,
  getTokenContractByAddress,
  getGovernorBravoDelegateContract,
  getVrtConverterProxyContract,
  getXvsVestingProxyContract,
  getVrtVaultProxyContract,
} from './getters';

export const useTokenContract = <T extends TokenId>(name: T) => {
  const web3 = useWeb3();
  return useMemo(() => getTokenContract<T>(name, web3), [web3, name]);
};

export const useTokenContractByAddress = (address: string) => {
  const web3 = useWeb3();
  return useMemo(() => getTokenContractByAddress(address, web3), [web3, address]);
};

export const useVTokenContract = <T extends VTokenId>(name: T) => {
  const web3 = useWeb3();
  return useMemo(() => getVTokenContract<T>(name, web3), [web3, name]);
};

export const useVaiUnitrollerContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getVaiUnitrollerContract(web3), [web3]);
};

export const useVaiVaultContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getVaiVaultContract(web3), [web3]);
};

export const useComptrollerContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getComptrollerContract(web3), [web3]);
};

export const usePriceOracleContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getPriceOracleContract(web3), [web3]);
};

export const useInterestModelContract = (address: string) => {
  const web3 = useWeb3();
  return useMemo(() => getInterestModelContract(address, web3), [web3]);
};

export const useVenusLensContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getVenusLensContract(web3), [web3]);
};

export const useXvsVaultContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getXvsVaultContract(web3), [web3]);
};

export const useXvsVaultProxyContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getXvsVaultProxyContract(web3), [web3]);
};

export const useGovernorBravoDelegateContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getGovernorBravoDelegateContract(web3), [web3]);
};

// VRT conversion
export const useVrtConverterProxyContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getVrtConverterProxyContract(web3), [web3]);
};

export const useXvsVestingProxyContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getXvsVestingProxyContract(web3), [web3]);
};

export const useVrtVaultProxyContract = () => {
  const web3 = useWeb3();
  return useMemo(() => getVrtVaultProxyContract(web3), [web3]);
};
