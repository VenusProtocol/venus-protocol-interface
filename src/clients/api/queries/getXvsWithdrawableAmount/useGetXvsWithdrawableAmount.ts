import { QueryObserverOptions, useQuery } from 'react-query';
import { callOrThrow } from 'utilities';

import getXvsWithdrawableAmount, {
  GetXvsWithdrawableAmountInput,
  GetXvsWithdrawableAmountOutput,
} from 'clients/api/queries/getXvsWithdrawableAmount';
import { DEFAULT_REFETCH_INTERVAL_MS } from 'constants/defaultRefetchInterval';
import FunctionKey from 'constants/functionKey';
import useGetUniqueContract from 'hooks/useGetUniqueContract';

type TrimmedGetXvsWithdrawableAmountInput = Omit<
  GetXvsWithdrawableAmountInput,
  'xvsVestingContract'
>;

type Options = QueryObserverOptions<
  GetXvsWithdrawableAmountOutput | undefined,
  Error,
  GetXvsWithdrawableAmountOutput | undefined,
  GetXvsWithdrawableAmountOutput | undefined,
  [FunctionKey.GET_XVS_WITHDRAWABLE_AMOUNT, TrimmedGetXvsWithdrawableAmountInput]
>;

const useGetXvsWithdrawableAmount = (
  input: TrimmedGetXvsWithdrawableAmountInput,
  options?: Options,
) => {
  const xvsVestingContract = useGetUniqueContract({
    name: 'xvsVesting',
  });

  return useQuery(
    [FunctionKey.GET_XVS_WITHDRAWABLE_AMOUNT, input],
    () =>
      callOrThrow({ xvsVestingContract }, params =>
        getXvsWithdrawableAmount({ ...params, ...input }),
      ),
    {
      refetchInterval: DEFAULT_REFETCH_INTERVAL_MS,
      ...options,
    },
  );
};

export default useGetXvsWithdrawableAmount;
