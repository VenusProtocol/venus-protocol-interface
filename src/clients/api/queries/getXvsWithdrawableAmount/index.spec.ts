import xvsVestingResponses from '__mocks__/contracts/xvsVesting';

import { XvsVesting } from 'packages/contracts';

import getXvsWithdrawableAmount from '.';

const fakeAccountAddress = '0x000000000000000000000000000000000AcCoUnt';

describe('api/queries/getXvsWithdrawableAmount', () => {
  test('returns the withdrawable amount on success', async () => {
    const xvsWithdrawableAmountMock = vi.fn(async () => xvsVestingResponses.withdrawableAmount);

    const fakeContract = {
      getWithdrawableAmount: xvsWithdrawableAmountMock,
    } as unknown as XvsVesting;

    const response = await getXvsWithdrawableAmount({
      xvsVestingContract: fakeContract,
      accountAddress: fakeAccountAddress,
    });

    expect(xvsWithdrawableAmountMock).toHaveBeenCalledTimes(1);
    expect(xvsWithdrawableAmountMock).toHaveBeenCalledWith(fakeAccountAddress);
    expect(response).toMatchSnapshot();
  });
});
