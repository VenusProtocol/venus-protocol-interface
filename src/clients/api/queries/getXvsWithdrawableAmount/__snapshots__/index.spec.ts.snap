// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/queries/getXvsWithdrawableAmount returns the withdrawable amount on success 1`] = `
Object {
  "totalVestedAmount": "1000",
  "totalWithdrawableAmount": "500000",
  "totalWithdrawnAmount": "0",
}
`;
